p2p/p2p-core/src/types.rs

```diff
@@ -0,0 +1,127 @@
+pub struct SetBan<A: NetZoneAddress> {
+    /// Address of the peer.
+    pub address: A,
+    /// - If [`Some`], how long this peer should be banned for
+    /// - If [`None`], the peer will be unbanned
+    pub ban: Option<Duration>,
+}
+
+/// Data within [`crate::services::AddressBookResponse::GetBans`].
+pub struct BanState<A: NetZoneAddress> {
+    /// Address of the peer.
+    pub address: A,
+    /// - If [`Some`], the peer is banned until this [`Instant`]
+    /// - If [`None`], the peer is not currently banned
+    pub unban_instant: Option<Instant>,
+}
+
+/// An enumeration of address types.
+///
+/// Used [`ConnectionInfo::address_type`].
+#[derive(Copy, Clone, Default, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[repr(u8)]
+pub enum AddressType {
+    #[default]
+    Invalid,
+    Ipv4,
+    Ipv6,
+    I2p,
+    Tor,
+}
+
+impl AddressType {
+    /// Convert [`Self`] to a [`u8`].
+    ///
+    /// ```rust
+    /// use cuprate_p2p_core::AddressType as A;
+    ///
+    /// assert_eq!(A::Invalid.to_u8(), 0);
+    /// assert_eq!(A::Ipv4.to_u8(), 1);
+    /// assert_eq!(A::Ipv6.to_u8(), 2);
+    /// assert_eq!(A::I2p.to_u8(), 3);
+    /// assert_eq!(A::Tor.to_u8(), 4);
+    /// ```
+    pub const fn to_u8(self) -> u8 {
+        self as u8
+    }
+
+    /// Convert a [`u8`] to a [`Self`].
+    ///
+    /// # Errors
+    /// This returns [`None`] if `u > 4`.
+    ///
+    /// ```rust
+    /// use cuprate_p2p_core::AddressType as A;
+    ///
+    /// assert_eq!(A::from_u8(0), Some(A::Invalid));
+    /// assert_eq!(A::from_u8(1), Some(A::Ipv4));
+    /// assert_eq!(A::from_u8(2), Some(A::Ipv6));
+    /// assert_eq!(A::from_u8(3), Some(A::I2p));
+    /// assert_eq!(A::from_u8(4), Some(A::Tor));
+    /// assert_eq!(A::from_u8(5), None);
+    /// ```
+    pub const fn from_u8(u: u8) -> Option<Self> {
+        Some(match u {
+            0 => Self::Invalid,
+            1 => Self::Ipv4,
+            2 => Self::Ipv6,
+            3 => Self::I2p,
+            4 => Self::Tor,
+            _ => return None,
+        })
+    }
+}
+
+// TODO: reduce fields and map to RPC type.
+//
+/// Data within [`crate::services::AddressBookResponse::ConnectionInfo`].
+pub struct ConnectionInfo<A: NetZoneAddress> {
+    pub address: A,
+    pub address_type: AddressType,
+    pub avg_download: u64,
+    pub avg_upload: u64,
+    pub connection_id: u64, // TODO: boost::uuids::uuid
+    pub current_download: u64,
+    pub current_upload: u64,
+    pub height: u64,
+    pub host: String,
+    pub incoming: bool,
+    pub ip: String,
+    pub live_time: u64,
+    pub localhost: bool,
+    pub local_ip: bool,
+    pub peer_id: String,
+    pub port: String,
+    pub pruning_seed: PruningSeed,
+    pub recv_count: u64,
+    pub recv_idle_time: u64,
+    pub rpc_credits_per_hash: u32,
+    pub rpc_port: u16,
+    pub send_count: u64,
+    pub send_idle_time: u64,
+    pub state: String, // TODO: what type is this?
+    pub support_flags: u32,
+}
```

Comment:
### TODO
- For `connection_id`, our currently used `u64` should be fine in place of `monerod`'s `boost::uuids::uuid`, right? The RPC output I think just needs a hex version of the ID.
- What can `{ip,port}` be replaced by `std::net::SocketAddr`? What about i2p/tor?
- What type is `{peer_id,state}`?

Context:
- https://github.com/monero-project/monero/blob/893916ad091a92e765ce3241b94e706ad012b62a/src/cryptonote_protocol/cryptonote_protocol_defs.h#L47
- https://www.getmonero.org/resources/developer-guides/daemon-rpc.html#get_connections

Reply:
@hinto-janai `boost::uuids::uuid` is 128 bit not 64 bit.
https://www.boost.org/doc/libs/1_86_0/libs/uuid/doc/html/uuid.html

Reply:
> For connection_id, our currently used u64 should be fine in place of monerod's boost::uuids::uuid, right? The RPC output I think just needs a hex version of the ID.

We just the peers address to identify the, we don't have a connection ID. If the address is unknown like an incoming Tor connection we use a u128.

> What can {ip,port} be replaced by std::net::SocketAddr? What about i2p/tor?

Yeah you should be able to pull them from the `std::net::SocketAddr`, if that is what you are asking. I don't know how monerod handles this for Tor.

> What type is {peer_id,state}?

peer_id is a u64 and state is a monerod specific string